_posts/2021-05-22-report-arms-spring2021.md

```diff
@@ -0,0 +1,101 @@
+
+With a grammar in hand, I used Lezer to generate a JavaScript parser for ARMS.  I then imported the parser into another JavaScript file and set it to parse user-entered input from a simple HTML interface.  One tricky area I ran into is that browser JavaScript currently offers poor support for import statements; to overcome this, I used [Webpack](https://webpack.js.org/) to bundle all of my code's dependencies into a single, self-contained file.
+
+![Bundled javascript...holy wall of text!](../assets/2021-05-22-report-arms-spring2021/bundled_javascript.png)  
+*The bundled javascript file for ARMS...holy wall of text!*
+
+I then wrote code to evaluate the Lezer-generated parse tree and generate the corresponding SDF code.  I gradually built up support for one SDF feature at a time, adding cubes, spheres, joints, and support for variable object scale and position.  As I went, the app grew to around 450 lines of JavaScript.
+
+Since text descriptions of 3D objects are error-prone due to being tough to mentally visualize, I decided to add an in-browser visualization of the user's objects.  I accomplished this with [ThreeJS](https://threejs.org/), adding code to dynamically generate and render a scene from the user's ARMS code as its parse tree is evaluated.
+
+## Results:
+
+![Current state of the ARMs app](../assets/2021-05-22-report-arms-spring2021/demo4.png)  
+*The ARMS app in action*
+
+The result is a simple web interface that lets the user enter short and intuitive ARMS code and click "compile".  The app will then output the corresponding, substantially longer SDF code below, along with a visualization of the objects created.
+
+Testing ARMS with inputs of a few different lengths saves the user, on average, **80%** by lines of code or **92%** by character count.  That is to say, using ARMS to generate a SDF simulation requires the user to type 92% fewer characters than writing the SDF by hand themselves.
+
+![Testing the ARMS app](../assets/2021-05-22-report-arms-spring2021/arms_table.png)
+
+*Three trials of ARMS with various length inputs*
+
+![ARMS graph](../assets/2021-05-22-report-arms-spring2021/arms_fig1.png)
+
+*Graph of the % of code saved based on input size*
+
+![ARMS graph 2](../assets/2021-05-22-report-arms-spring2021/arms_fig2.png)
+
+*The savings are even higher when considered by character count.*
+
+The source code for the app can be found [on my GitHub](https://github.com/swow2015/arms2).  The app can be quickly demoed by downloading ["dist/main.js"](https://github.com/oapostrophe/arms2/blob/main/dist/main.js), ["dist/app.html"](https://github.com/oapostrophe/arms2/edit/main/dist/app.html), and ["dist/style.css"](https://github.com/oapostrophe/arms2/blob/main/dist/style.css).  With these three in the same directory, you can simply open the HTML file and the app will run.
```

Comment:
Will you eventually push the code to https://github.com/ARMS/arms.github.io?

Reply:
I can if desired, previously I remember we had decided to make a fresh repository for the new code.

Reply:
I happen to like that URL, but I'm open to suggestions.

Reply:
I'll add the new code and give it a push this evening!